docs/apis-tools/frontend-development/03-forms/02-embed-forms/02-embed-in-javascript.md

```diff
@@ -0,0 +1,245 @@
+} } />
+
+### Validate a form
+
+Before you allow a user to submit a form, you can use the `validate` function to ensure that all validation rules of your form are met and that all required fields are completed. Learn more in the [form API documentation](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#formvalidate--errors) <GHIcon />.
+
+```js
+const errors = form.validate();
+
+if (Object.keys(errors).length) {
+  console.error('Form has errors', errors);
+}
+```
+
+### Trigger and listen to form events
+
+Form-js provides a comprehensive set of events and APIs to react on form state changes. You can listen for 
+
+- [form state changes](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#changed---data-errors-) <GHIcon />,
+- [form submissions](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#submit---data-errors-) <GHIcon />,
+- [form layout changes](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#layouting-events) <GHIcon />.
+
+In addition, hook into [lifecycle events](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#lifecycle-events) <GHIcon /> to add custom logic (e.g. initialize listeners on form fields after the form loaded).
+
+Learn more about the full API in the [GitHub repository documentation](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#api) <GHIcon />.
+
+### Retrieve form output data
+
+To retrieve the current form output data on any form state change, listen to the [`changed`](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#changed---data-errors--) <GHIcon /> event.
+
+To retrieve the data on submit, listen to the [`submit`](https://github.com/bpmn-io/form-js/tree/develop/packages/form-js-viewer#formsubmit---data-data-errors-errors-) <GHIcon /> event.
+
+### Retrieve schema variables
+
+Use the `getSchemaVariables` util to retrieve the [variables defined in a form schema](./01-concepts.md#schema-variables). This is useful to gather what data is consumed and produced by a form.
+
+```javascript
+import { getSchemaVariables } from '@bpmn-io/form-js';
+
+const variables = getSchemaVariables(schema);
+
+console.log('Schema variables', variables);
+```
+
+It is also possible to distinct between input and output variables:
+
+```javascript
+import { getSchemaVariables } from '@bpmn-io/form-js';
+
+const outputVariables = getSchemaVariables(schema, { inputs: false });
+const inputVariables = getSchemaVariables(schema, { outputs: false });
+```
+
+:::note
+form-js does not enforce typing. Retrieving schema variables returns the variable names, but not the type or whether the variable is optional (i.e. whether the field is required or not). To retrieve the expected type of the variable, parse the form schema manually. To enforce the typing of input variables, use validation libraries such as [joi](https://github.com/hapijs/joi) <GHIcon />.
```

Comment:
Do we have the required stuff as a feature request? Might be worth raising, I know types are already requested but required may not, and it's easy to do :).

Reply:
Not to my knowledge. Users ask for typing on the task level, not on the form level (so far)